R/reportAgg.R

```diff
@@ -0,0 +1,287 @@
+#' @returns named list with definition of common set elements
+#'
+#' @importFrom madrat toolGetMapping
+#' @importFrom yaml read_yaml
+
+.readBrickSets <- function(res) {
+
+  readIt <- function(file) {
+    toolGetMapping(name = file,
+                   type  = "sectoral",
+                   where = "reportBrick",
+                   returnPathOnly = TRUE) %>%
+      read_yaml()
+  }
+
+  file <- "brickSets.yaml"
+  brickSets <- readIt(file)
+
+  # replace default sets with custom sets where defined
+  if (!is.null(res)) {
+    file <- paste0("brickSets_", res, ".yaml")
+    customBrickSets <- readIt(file)
+    brickSets[names(customBrickSets)] <- customBrickSets
+  }
+
+  # duplicate aliases
+  brickSetsExplicit <- list()
+  for (dimName in names(brickSets)) {
+    dim <- brickSets[dimName]
+    aliases <- dim[[1]][["alias"]]
+    dim[[1]][["alias"]] <- NULL
+    aliasDims <- rep(dim, length(aliases))
+    names(aliasDims) <- aliases
+    brickSetsExplicit <- c(brickSetsExplicit, c(as.list(dim), aliasDims))
+  }
+
+  attr(brickSetsExplicit, "file") <- file
+  return(brickSetsExplicit)
+}
+
+
+
+
+
+#' Create tag
+#'
+#' Escape dimension name in curly brackets
+#'
+#' @param dim character dimension name
+#' @returns character, dimension tag
+
+.tag <- function(dim) {
+  paste0("{", dim, "}")
+}
+
+
+
+
+
+#' Aggregate
+#'
+#' @param x MagPIE object, BRICK object
+#' @param agg named vector of dimensions to aggregate.
+#' @returns aggregated MagPIE objects without sub dimensions in dim 3
+#'
+#' @importFrom magclass dimSums mselect
+
+.agg <- function(x, agg) {
+
+  # return NULL if any element in any dimension is missing
+  missingElements <- .missingElements(x, agg)
+  if (length(missingElements) > 0) {
+    return(NULL)
+  }
+
+  do.call(mselect, c(list(x = x, agg))) %>%
+    dimSums(na.rm = TRUE)
+}
+
+
+
+
+
+#' List any missing elements in any given dimension
+#'
+#' @param x MagPIE object, BRICK object
+#' @param dimLst named vector, dimension ans names and elements as values
+#' @returns vector of missing dimension elements in x
+#'
+#' @importFrom magclass getItems
+
+.missingElements <- function(x, dimLst) {
+  missingDims <- setdiff(names(dimLst), getSets(x))
+  if (length(missingDims) > 0) {
+    stop("The following dimensions are listed in 'dimLst' but missing in 'x': ",
+         paste(missingDims, collapse = ", "))
+  }
+  unlist(lapply(names(dimLst), function(dim) {
+    if (!dim %in% getSets(x)) {
+      stop("x has no dimension call")
+    }
+    setdiff(dimLst[[dim]], getItems(x, dim = dim))
+  }))
+}
+
+
+
+
+
+#' Set names is not NULL
+#'
+#' @param object MAgPIE object
+#' @param nm a vector of names current names should be replaced with. If only
+#'   one data element exists you can also set the name to NULL.
+#' @returns NULL if object is NULL else object with manipulated names
+#'
+#' @importFrom magclass setNames
+
+.setNames <- function(object, nm) {
+  if (is.null(object)) return(NULL)
+  setNames(object, nm)
+}
```

Comment:
I adapted it lightly to have the early return on an individual line. Else I followed the tidyverse coding style advice to avoid the final return statement. Though I'm not sure I would agree with this recommendation. I'M also fine with explicit returns.